templates/test/dev/custom-builds/patches/custom-builds.yaml

```diff
@@ -8,73 +8,50 @@ spec:
   kubeadmConfigSpec:
     useExperimentalRetryJoin: true
     clusterConfiguration:
-      kubernetesVersion: "${KUBERNETES_VERSION}"
+      kubernetesVersion: "ci/${CI_VERSION}"
+    preKubeadmCommands:
+    - bash -c /tmp/replace-k8s-binaries.sh
     postKubeadmCommands:
     - bash -c /tmp/replace-k8s-components.sh
     files:
+    - path: /tmp/replace-k8s-binaries.sh
+      owner: "root:root"
+      permissions: "0744"
+      content: |
+        #!/bin/bash
+
+        set -o nounset
+        set -o pipefail
+        set -o errexit
+
+        declare -a BINARIES=("kubeadm" "kubectl" "kubelet")
+        for BINARY in "$${BINARIES[@]}"; do
+          echo "* installing package: $${BINARY} ${KUBE_GIT_VERSION}"
+          curl --retry 10 --retry-delay 5 "https://${AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/${JOB_NAME}/${KUBE_GIT_VERSION}/bin/linux/amd64/$${BINARY}" --output "/usr/bin/$${BINARY}"
+        done
+        systemctl restart kubelet
+
+        echo "kubeadm version: $(kubeadm version -o=short)"
+        echo "kubectl version: $(kubectl version --client=true --short=true)"
+        echo "kubelet version: $(kubelet --version)"
```

Comment:
Moved binary extraction to `preKubeadmCommands` so that the version skew between the kubeadm binary and
the control plane that it is going to initialize is minimized. For example, using kubeadm binary built from HEAD to initialize a v1.22.0-alpha.0 control plane is much less error-prone than using kubeadm v1.18.8. 
 
Also removed retry helper functions in favor of `curl` built-in `--retry` and `--retry-delay`. 

/cc jackfrancis

Reply:
Everything about this changeset looks awesome